stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -260,11 +303,21 @@ public struct Unsafe${Mutable}BufferPointer<Element>
         base: self, bounds: bounds)
     }
 %  if Mutable:
-    set {
+    nonmutating set {
       _debugPrecondition(bounds.lowerBound >= startIndex)
       _debugPrecondition(bounds.upperBound <= endIndex)
+      _debugPrecondition(bounds.count == newValue.count)
+
       // FIXME: swift-3-indexing-model: tests.
-      _writeBackMutableSlice(&self, bounds: bounds, slice: newValue)
+      if newValue.count > 0 {
+        var sliceIndex = newValue.startIndex
+        var index = bounds.lowerBound
+        while index != bounds.upperBound && sliceIndex != newValue.endIndex {
```

Comment:
Given we know these are two unsafe buffers, we ought to just to a straight memory copy, once we've checked the count is the same. Something like...

```
  baseAddress!.initialize(from: newValue.base.baseAddress!, count: bounds.count)
```

The only thing that would worry me about if this is correct is the possibility of overlapping assignment from `self`...

Reply:
`UnsafeMutablePointer.assign` supports overlapping ranges. That's the thing to use.

Reply:
Yes, exactly but use assignment.
`baseAddress!.assign(from: newValue.base.baseAddress!, count: bounds.count)`

That does the correct forward or backward copy depending on the relative pointer positions.